services/device/geolocation/geoclue_location_provider.cc

```diff
@@ -0,0 +1,414 @@
+  }
+
+  gclue_client_ = bus_->GetObjectProxy(kServiceName, path);
+  SetClientProperties();
+}
+
+void GeoClueLocationProvider::SetClientProperties() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  auto* raw_ptr = new GeoClueProperties(gclue_client_.get());
+  std::unique_ptr<dbus::PropertySet> properties(raw_ptr);
+
+  // Wait for all properties to be set before starting the client.
+  const auto callback = base::BarrierCallback<bool>(
+      2, base::BindOnce(&GeoClueLocationProvider::OnSetClientProperties,
+                        weak_ptr_factory_.GetWeakPtr(), std::move(properties)));
+
+  raw_ptr->requested_accuracy_level.Set(
+      high_accuracy_requested_ ? kGeoClueAccuracyHigh : kGeoClueAccuracyLow,
+      callback);
+
+  raw_ptr->desktop_id.Set(kBraveDesktopId, callback);
+}
+void GeoClueLocationProvider::OnSetClientProperties(
+    std::unique_ptr<dbus::PropertySet> property_set,
+    std::vector<bool> success) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  if (base::ranges::any_of(success, [](const auto& value) { return !value; })) {
+    LOG(ERROR) << "Failed to set properties. GeoClue2 location provider will "
+                  "not work properly";
+    SetPosition(GetErrorPosition());
+    return;
+  }
+
+  ConnectSignal();
+}
+
+void GeoClueLocationProvider::ConnectSignal() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  gclue_client_->ConnectToSignal(
+      kClientInterfaceName, "LocationUpdated",
+      base::BindRepeating(
+          [](base::WeakPtr<GeoClueLocationProvider> provider,
+             dbus::Signal* signal) {
+            dbus::MessageReader reader(signal);
+            dbus::ObjectPath old_location;
+            dbus::ObjectPath new_location;
+            if (!reader.PopObjectPath(&old_location) ||
+                !reader.PopObjectPath(&new_location)) {
+              if (provider) {
+                provider->SetPosition(GetErrorPosition());
+              }
+              return;
+            }
+
+            if (provider) {
+              provider->ReadGeoClueLocation(new_location);
+            }
+          },
+          weak_ptr_factory_.GetWeakPtr()),
+      base::BindOnce(&GeoClueLocationProvider::OnSignalConnected,
+                     weak_ptr_factory_.GetWeakPtr()));
+}
+
+void GeoClueLocationProvider::OnSignalConnected(
+    const std::string& interface_name,
+    const std::string& signal_name,
+    bool success) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  if (!success) {
+    LOG(ERROR) << "Failed to connect to LocationUpdated Signal. GeoClue2 "
+                  "location provider will "
+                  "not work properly";
+    SetPosition(GetErrorPosition());
+    return;
+  }
+
+  client_state_ = kInitialized;
+  StartClient();
+}
+
+void GeoClueLocationProvider::StartClient() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  if (!gclue_client_ || !permission_granted_ || client_state_ != kInitialized) {
```

Comment:
https://github.com/brave/brave-core/pull/18063/files#diff-29866bd53035388e9e1965be15699a6a8d75735aa4297124b601fd2a4c393871R95 says that it's ok to call `StartClient` before permission is granted?